### PR TITLE
[Close #42] Euen.42.update to cowboy 2

### LINKED ISSUE #42: Update Lasse in order to work fine with Cowboy 2.x

### PR DESCRIPTION
**Pull request by Euen**
[Close #42] Euen.42.update to cowboy 2

### CODE REVIEW

File: examples/ping_pong/Makefile [hunk 1]
```diff
@@ -2,11 +2,11 @@ PROJECT = ping_pong
 
 DEPS = cowboy lasse sync
 
-dep_cowboy = https://github.com/extend/cowboy.git 0.10.0
-dep_lasse = ../../ master
-dep_sync = https://github.com/rustyio/sync master
+dep_cowboy = git https://github.com/ninenines/cowboy.git  2.0.0-pre.3
+dep_lasse = git https://github.com/inaka/lasse.git davecaos.cowboy2x
```
Comment (elbrujohalcon) [inline examples/ping_pong/Makefile]: Please replace this branch by `cowboy2`. Even if it doesn't work now, it will work once this PR is merged.
`davecaos.cowboy2x` branch will probably be deleted soon.

File: examples/ping_pong/src/ping_pong_sup.erl [hunk 2]
```diff
@@ -16,8 +16,8 @@ start_link() ->
     supervisor:start_link(?MODULE, {}).
 
 start_listeners() ->
-    Port = 8080,
-    ListenerCount = 1,
+    Port          = application:get_env(nss, http_port, 8080),
```
Comment (elbrujohalcon) [inline examples/ping_pong/src/ping_pong_sup.erl]: why have you changed this? `nss` is not an app that will run here, so it will end up as 8080 (and next line as 1) always. And it's an example app, so it really doesn't matter. Please undo this change.

File: src/lasse_handler.erl [hunk 3]
```diff
@@ -20,186 +21,190 @@
...
+  ModuleState = State#state.state,
+  Module:terminate(Reason, Req, ModuleState),
+  ok.
+
+-spec upgrade(Req, Env, module(), any(), timeout(), run | hibernate)
+  -> {ok, Req, Env} | {suspend, module(), atom(), [any()]}
+  when Req::cowboy_req:req(), Env::cowboy_middleware:env().
+upgrade(Req, Env, Handler, HandlerState, Timeout, run) ->
+  cowboy_loop:upgrade(Req, Env, Handler, HandlerState, Timeout, run);
+upgrade(Req, Env, Handler, HandlerState, Timeout, hibernate) ->
```
Comment (elbrujohalcon) [inline src/lasse_handler.erl]: Why 2 clauses if you're basically doing the same in both? Why not just:
```erlang
upgrade(Req, Env, Handler, HandlerState, Timeout, Action) ->
  cowboy_loop:upgrade(Req, Env, Handler, HandlerState, Timeout, Action).
```

you could've even just _mixed_ the function _in_ using mixer.

File: src/lasse_handler.erl [hunk 4]
```diff
@@ -20,186 +21,190 @@
...
+        fun(Event) -> ok = send_event(Event, Req1) end,
+        InitialEvents
+      ),
+      {cowboy_loop, Req1, #state{module = Module, state = State}};
+
+   _OtherMethod ->
+      Headers = [{<<"content-type">>, <<"text/html">>}],
+      StatusCode = 405, % Method not Allowed
+      Req1 = cowboy_req:reply(StatusCode, Headers, Req),
+      {cowboy_loop, Req1, #state{module = Module}}
```
Comment (elbrujohalcon) [inline src/lasse_handler.erl]: You replaced `shutdown` by `cowboy_loop`. Is that expected, how are you telling cowboy to just return to the client without actually looping? Is that accomplished using `cowboy_req:reply/3` in the line above?

File: src/lasse_handler.erl [hunk 5]
```diff
@@ -20,186 +21,190 @@
...
-            ModuleState = State#state.state,
-            ErrorNewState = Module:handle_error(Event, Reason, ModuleState),
-            {ok, Req, State#state{state = ErrorNewState}};
-        ok ->
-            {loop, Req, State#state{state = NewState}}
-    end;
+  try send_event(Event, Req) of
+    ok -> {ok, Req, State#state{state = NewState}}
+  catch
+    Reason ->
```
Comment (elbrujohalcon) [inline src/lasse_handler.erl]: you should use `_:Reason ->` here.

File: test/events_handler.erl [hunk 6]
```diff
@@ -10,18 +10,15 @@
...
-            ct:pal("Initiating a ~p in ~p", [ProcName, whereis(ProcName)]);
-        {undefined, Req}  ->
-            ct:pal("Initiating handler"),
-            ok
-    end,
-
-    {ok, Req, LastEventId}.
+  % Take process name from the "process-name" header.
+  case cowboy_req:header(<<"process-name">>, Req) of
+      ProcNameBin when ProcNameBin =/= <<"undefined">> ->
```
Comment (elbrujohalcon) [inline test/events_handler.erl]: we use 2 spaces for indentation ;)

File: test/lasse_handler_SUITE.erl [hunk 7]
```diff
@@ -169,16 +177,18 @@ send_last_event_id(_Config) ->
...
-        State = {state, events_handler, {}},
-
-        meck:new(cowboy_req, [passthrough]),
-        meck:expect(cowboy_req, chunk, fun(_, _) -> {error, just_because} end),
-        {ok, Req, _}  = lasse_handler:info(send, Req, State)
-    after
-        catch meck:unload(cowboy_req)
-    end.
+  ok = try
+          Req = {},
```
Comment (elbrujohalcon) [inline test/lasse_handler_SUITE.erl]: bad indentation

File: test/lasse_handler_SUITE.erl [hunk 8]
```diff
@@ -196,14 +206,14 @@ shutdown(_Config) ->
 init_without_module_option(_Config) ->
     ok = try
              Opts = [],
-             lasse_handler:init({}, {}, Opts),
+             lasse_handler:init({}, Opts),
```
Comment (elbrujohalcon) [inline test/lasse_handler_SUITE.erl]: should have 3 arguments.

File: test/lasse_handler_SUITE.erl [hunk 9]
```diff
@@ -196,14 +206,14 @@ shutdown(_Config) ->
...
              fail
          catch
              throw:module_option_missing -> ok
          end,
     ok = try
              Opts2 = #{init_args => []},
-             lasse_handler:init({}, {}, Opts2),
+             lasse_handler:init({}, Opts2),
```
Comment (elbrujohalcon) [inline test/lasse_handler_SUITE.erl]: should have 3 arguments.

File: test/shutdown_handler.erl [hunk 10]
```diff
@@ -11,10 +11,18 @@
          terminate/3
         ]).
 
+-record(state,
```
Comment (elbrujohalcon) [inline test/shutdown_handler.erl]: why do you need this?

File: src/lasse_handler.erl [hunk 11]
```diff
@@ -1,8 +1,10 @@
 %%% @doc Server-Sent Event handler for Cowboy
 -module(lasse_handler).
 
+-behaviour(cowboy_loop).
```
Comment (gadgetci) [inline src/lasse_handler.erl]: According to **Compiler**:
> Warning: behaviour cowboy_loop undefined

Comment (gadgetci) [inline src/lasse_handler.erl]: According to **Dialyzer**:
>  Callback info about the cowboy_loop behaviour is not available

File: src/lasse_handler.erl [hunk 12]
```diff
@@ -20,186 +22,182 @@
...
-    catch
-      _:{badmatch, #{}} ->
-        throw(module_option_missing)
-    end.
+  module() |
+  #{ module => module()
+   , init_args => any()
+   }.
+
+-spec init(cowboy_req:req(), lasse_handler_options()) ->
```
Comment (gadgetci) [inline src/lasse_handler.erl]: According to **Dialyzer**:
>  The return type {atom() | tuple(),_,#state{module:atom() | tuple()}} in the specification of init/2 is not a subtype of {atom() | tuple(),cowboy_req:req(),_} | {atom() | tuple(),cowboy_req:req(),_,'hibernate' | 'infinity' | non_neg_integer()} | {atom() | tuple(),cowboy_req:req(),_,'infinity' | non_neg_integer(),'hibernate'}, which is the expected return type for the callback of cowboy_loop behaviour

File: src/lasse_handler.erl (hunk 11 above)
Comment (gadgetci) [inline src/lasse_handler.erl]: According to **Compiler**:
> Warning: behaviour cowboy_loop undefined

File: Makefile [hunk 13]
```diff
@@ -1,14 +1,15 @@
 PROJECT = lasse
 
+DEPS = cowboy
```
Comment (elbrujohalcon) [inline Makefile]: Don't you need to add this as well to rebar.config ?

File: src/lasse_handler.erl (hunk 11 above)
Comment (gadgetci) [inline src/lasse_handler.erl]: According to **Compiler**:
> Warning: behaviour cowboy_loop undefined
